backend/src/gateway/index.ts

```diff
@@ -1,37 +1,71 @@
 import Messaging from "../models/Messages";
 import { Server } from "socket.io";
 import { Server as HTTPServer } from "http";
+import GroupChats from "../models/Groupchat";
 
 class Gateway {
-	socketIO: Server;
-	constructor(http: HTTPServer) {
-		this.socketIO = new Server(http);
-		console.log("Gateway initialized");
-	}
-
-	init() {
-		this.socketIO.on("connection", (socket) => {
-			console.log(`SOCKETIO: ${socket.id} user just connected!`);
-			Messaging.find().then((result) => {
-				socket.emit("output-messages", result);
-			});
-
-			socket.on("message", this.messageHandler.bind(this, socket));
-
-			socket.on("disconnect", () => {
-				console.log(`${socket.id} user disconnected`);
-			});
-		});
-	}
-
-	messageHandler(socket, msg) {
-		const message = new Messaging({ msg });
-		message.save().then((document) => {
-			console.log(document)
-			console.log("SAVING MESSAGE \"" + msg +"\" from user " + socket.id);
-			this.socketIO.emit("message", msg);
-		});
-	}
+  socketIO: Server;
+  constructor(http: HTTPServer) {
+    this.socketIO = new Server(http);
+    console.log("Gateway initialized");
+  }
+
+  init() {
+    this.socketIO.use((socket, next) => {
+      const username = socket.handshake.auth.username;
+      if (!username) {
+        return next(new Error("invalid username"));
+      }
+      (socket as any).username = username;
+      next();
+    });
+
+    this.socketIO.on("connection", (socket) => {
+      console.log(`SOCKETIO: ${socket.id} user just connected!`);
+
+      socket.on("send_message", this.onSendMessage.bind(this, socket));
+
+      Messaging.find().then((result) => {
+        socket.emit("output-messages", result);
+      });
+
+      socket.on("message", this.messageHandler.bind(this, socket));
+
+      socket.on("disconnect", () => {
+        console.log(`${socket.id} user disconnected`);
+      });
+    });
+  }
+
+  async onSendMessage(socket, content, group_id, userName) {
+    try {
+      const date = Date.now()
+      const messageToBeAdded = new Messaging({
+        message: content,
+        groupChat: group_id,
+        user: userName,
+        timestamp: date
+      });
+      const added = await messageToBeAdded.save();
+      // emit to all of the ids in the group
+      const groupchat = await GroupChats.findById({ _id: { group_id } });
+      for (let i = 0; i < groupchat.users.length; i++) {
+        socket.to(groupchat.users[i]).emit("send_message", { added });
```

Comment:
https://stackoverflow.com/questions/48811315/passport-socketio-how-to-get-a-list-of-online-users-with-nodejs-express-and-p

Implement this to get all the online users (should be very quick new file in gateway, call it ActiveUsers.js and it basically just maintains a set of all online users and returns whether a user is online or not, and if so then give the socketio id of that user